themes/default/content/blog/pulumi-is-imperative-declarative-imperative/index.md

```diff
@@ -0,0 +1,221 @@
+---
+title: "Pulumi Is Imperative, Declarative and Imperative"
+date: 2022-09-30T12:28:19+02:00
+meta_desc: Pulumi is often debated to be imperative or declarative.
+  What if I tell you that Pulumi is imperative, declarative and imperative?
+meta_image: meta.png
+authors:
+    - ringo-de-smet
+tags:
+    - cloud-engineering
+    - infrastructure-as-code
+---
+
+On a regular basis, articles and tweets pass by discussing whether or not some specific tool is imperative or declarative.
+It's no surprise that Pulumi is often the tool being debated. What if I tell you that Pulumi is imperative, declarative and imperative?
+
+<!--more-->
+
+When we look at our [Frequently Asked Questions](/docs/support/faq/#is-pulumi-imperative-or-declarative), we read the following on the declarative versus imperative topic:
+
+> Pulumi is a declarative tool that uses imperative languages to define your end state. The language is used for authoring your program.
+> It’s not used for talking to the cloud provider API.
+
+It is good to know to refresh what `declarative` and `imperative` mean:
```

Comment:
Also, I did a Quick Bites on imperative vs declarative infrastructure that got into this confusion (and how a tool can be both): https://www.youtube.com/watch?v=4B7vLfVYG84. It may be worth linking somewhere here.

Reply:
As the outcome of the conversation we have over Slack, I'll update the headings of the 3 Pulumi architectural components as you proposed:

* "An Imperative Part: The Language Host"
* "The Declarative Part: The CLI and Engine"
* "Another Imperative Part: The Providers"

Makes it indeed clearer why I titled the article: `Pulumi is imperative, declarative, and imperative`

Thanks for the help!